### PR TITLE
StaticHtmlReporter: Break down the issue summary

### PR DESCRIPTION
**Pull request by sschuberth**
Display the numbers of errors, warnings and hints separately.

Signed-off-by: Sebastian Schuberth <sebastian.schuberth@here.com>

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1331)

### CODE REVIEW

File: reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt [hunk 1]
```diff
@@ -379,13 +379,58 @@ class StaticHtmlReporter : Reporter() {
...
+                issueRow.scanIssues.forEach { (_, issues) ->
+                    issues.forEach {
+                        when (it.severity) {
+                            Severity.ERROR -> ++errorCount
+                            Severity.WARNING -> ++warningCount
+                            Severity.HINT -> ++hintCount
+                        }
+                    }
+                }
+            }
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt]: Same here, e.g.:

```kotlin
val allIssues = reportTableModel.issueSummary.rows.flatMap {
    it.analyzerIssues.flatMap { (_, issues) -> issues } + it.scanIssues.flatMap { (_, issues) -> issues }
}
val errorCount = allIssues.count { it.severity == Severity.ERROR }
```

Adding these error counts to the `IssueRow` class would keep the presentation code here cleaner and make the data available to other table reports.

File: reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt [hunk 2]
```diff
@@ -379,13 +379,58 @@ class StaticHtmlReporter : Reporter() {
...
+                var warningCount = 0
+                var hintCount = 0
+
+                issues.forEach {
+                    when (it.severity) {
+                        Severity.ERROR -> ++errorCount
+                        Severity.WARNING -> ++warningCount
+                        Severity.HINT -> ++hintCount
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt]: I indeed had it like that before, but then (even if not performance critical) found it inelegant to iterate over the list of issues three times, once per severity (and similar below).

File: reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt (hunk 1 above)
Comment (sschuberth) [reply reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt]: I'll look into adding these numbers to `IssueRow`.

File: reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt [hunk 3]
```diff
@@ -379,13 +379,27 @@ class StaticHtmlReporter : Reporter() {
         h2 { +"Index" }
 
         ul {
-            reportTableModel.evaluatorIssues?.let {
-                li { a("#policy-violation-summary") { +"Rule Violation Summary (${it.size} violations)" } }
+            reportTableModel.evaluatorIssues?.let { ruleViolations ->
+                val issues = ruleViolations.groupBy { it.severity }
+                val errorCount = issues[Severity.ERROR].orEmpty().count()
+                val warningCount = issues[Severity.WARNING].orEmpty().count()
+                val hintCount = issues[Severity.HINT].orEmpty().count()
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt]: There is also this option:

```kotlin
        val counts = issues.groupingBy { it.severity }.eachCount()
        val errors = counts[Severity.ERROR] ?: 0
```

I was curious which performs better and found that the variant with `groupBy` is slightly faster up to 1,000,000 elements in the list, but above that the performance breaks down and at 10,000,000 elements it is 20 times slower than the variant with `groupingBy`.

Comment (sschuberth) [reply reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt]: Interesting! Assuming that we mostly have < 1,000,000 issues, I guess we stick to the current solution 😉

Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt]: Yes, I think if you have so many issues you can wait a few more seconds for the report ;-)
Btw, the solution you had in the first version of the PR is 2x faster for a reasonable number of issues (<1000), but we are talking about <1ms here.

Comment (sschuberth) [reply reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt]: Thanks for taking the time to benchmark that one, too. Good to know I can still write performant code 😆

**Pull request merged**